### PR TITLE
Support UVerb in servant-auth-server

### LINKED ISSUE #1570: UVerb HasServer Instance
Been trying to debug this all day but I think it's not working because maybe there's an interface missing for servant-auth with UVerbs.  Maybe that's not supported yet?

The error I'm getting is as follows.
```
/home/robwithhair/GGMR/vfde-inv-var-api/src/API.hs:112:55: error:
    • No instance for (HasServer  
                         (Servant.Auth.Server.Internal.AddSetCookie.AddSetCookieApi
                            (Servant.Auth.Server.Internal.AddSetCookie.AddSetCookieApi
                               (UVerb
                                  'PUT
                                  '[JSON]
                                  '[WithStatus 201 NoContent, WithStatus 204 NoContent])))
                         '[JWTSettings, CookieSettings,
                           BasicAuthData -> IO (AuthResult AuthenticatedUser)])
        arising from a use of ‘serveWithContext’
    • In the second argument of ‘($)’, namely
        ‘serveWithContext api cfg (server config)’
      In a stmt of a 'do' block:  
        pure . ifRequest shouldApplyApiGateway apiGateway
          $ serveWithContext api cfg (server config)
      In the expression:          
        do myKey <- fromSecret <$> ByteString.readFile jwtSecretPath
           let jwtCfg = defaultJWTSettings myKey
               authCfg = authCheck config
               ....               
           pure . ifRequest shouldApplyApiGateway apiGateway
             $ serveWithContext api cfg (server config)
    |                             
112 |   pure . ifRequest shouldApplyApiGateway apiGateway $ serveWithContext api cfg (server config)
    |                                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                                  
```

Not sure what I'm missing here, so would be infinitely grateful for any input.

```
type LocallyStoredItemAPI (sectionName :: Symbol) listType detailType patchType patchResponseType = sectionName :>
                                            ( Get '[JSON] listType    -- LIST
                                              :<|> Capture "id" ItemID :> Get '[JSON] detailType  -- GET
                                              :<|> Capture "id" ItemID :> ReqBody '[JSON] detailType :> UVerb 'PUT '[JSON] '[WithStatus 201 NoContent, WithStatus 204 NoContent]  -- PUT
                                              :<|> Capture "id" ItemID :> ReqBody '[JSON] patchType :> Patch '[JSON] patchResponseType -- PATCH
                                              :<|> ReqBody '[JSON] detailType :> PostCreated '[JSON] NoContent -- POST
                                              :<|> Capture "id" ItemID :> Delete '[JSON] NoContent -- DELETE
                                           )

putLocallyStoredItem :: (SetModifiedTime i, ToJSON i)
  => Config
  -> (LocallyStoredItemId -> C.RelativePath)
  -> LocallyStoredItemId
  -> i
  -> Handler (Union '[WithStatus 201 NoContent, WithStatus 204 NoContent])
putLocallyStoredItem cfg dsJsonFileFromId dsid body = do
  itemExistsBefore <- liftIO $ (C.fileExistsF cfg) (dsJsonFileFromId dsid)
  -- Write to disk with modified time
  liftIO $ withCurrentModifiedTime body >>= writeDsFile . encode
  if itemExistsBefore then
    Servant.Server.UVerb.respond (WithStatus @204 NoContent)
  else
    Servant.Server.UVerb.respond (WithStatus @201 NoContent)
  where writeDsFile = C.writeFileF cfg (dsJsonFileFromId dsid)
```
I realise this isn't a very minimal example.  If I can I'll work on one but on a deadline ATM.  Hope you understand.  I'm following the examples from servant Cookbook.  Looked everywhere for an import to add the interfaces but think writing them myself may be beyond my understanding of Servant ATM.

### PR DESCRIPTION
**Pull request by gdeest**
Closes #1570 

~Caveat: returning a 401 response with UVerb is _not_ equivalent to using `throwAll err401`: `XSRF-TOKEN` will be set anyway, because the handler then doesn't fail (it successfully returns an error :upside_down_face:). I can't see a satisfying solution to this problem, so external input is much welcome on this issue (I for one consider this limitation semi-reasonable).~ This should be fixed.

### CODE REVIEW

File: changelog.d/1571 [hunk 1]
```diff
@@ -0,0 +1,8 @@
+synopsis: Support UVerb in servant-auth-server
+prs: #1571
+issues: #1570
+description: {
+UVerb endpoints are now supported by servant-auth-server and can used under the
```
Comment (googleson78) [inline changelog.d/1571:5]:
```suggestion
UVerb endpoints are now supported by servant-auth-server and can be used under the
```

File: servant-auth/servant-auth-server/src/Servant/Auth/Server/Internal/AddSetCookie.hs [hunk 2]
```diff
@@ -33,12 +34,24 @@ type family AddSetCookieApiVerb a where
...
+
 type family AddSetCookieApi a :: *
 type instance AddSetCookieApi (a :> b) = a :> AddSetCookieApi b
 type instance AddSetCookieApi (a :<|> b) = AddSetCookieApi a :<|> AddSetCookieApi b
+#if MIN_VERSION_servant_server(0,19,0)
 type instance AddSetCookieApi (NamedRoutes api) = AddSetCookieApi (ToServantApi api)
+#endif
 type instance AddSetCookieApi (Verb method stat ctyps a)
   = Verb method stat ctyps (AddSetCookieApiVerb a)
+#if MIN_VERSION_servant_server(0,18,1)
```
Comment (gdeest) [inline servant-auth/servant-auth-server/src/Servant/Auth/Server/Internal/AddSetCookie.hs:51]: Note to self: this is in fact not sufficient, as `UVerb` support depends on some instances in `servant` that are provided by this PR and hence not yet in any released version of `servant`.

**Comment by tchoutri** [comment]
@gdeest Is this ready for review? :)

**Comment by gdeest** [comment]
@tchoutri Modulo the small comments for myself, yes, it is !

**Comment by RileyEv** [comment]
@gdeest @tchoutri Any chance of getting this merged? I'd really appreciate it 😄

**Comment by tchoutri** [comment]
No idea how I missed that one, sorry!

**Comment by robwithhair** [comment]
@tchoutri would you mind recreating the branch servant-auth-uverb?  Unfortunately due to the way GitHub squashes commits on merged pull requests this has broken my ability to pull this commit in stack and thus my code won't build any longer.  I tried building my code on master but there have been several changes and several dependencies since then.  I can fork but not recreate the branch once forked.

**Comment by RileyEv** [comment]
The branch still exists on my fork, which I made while waiting for this to be merged, feel free to take from there, but don't rely on it for long as I might delete the fork now that this has been merged

https://github.com/bank-manager/servant/tree/servant-auth-uverb

**Pull request merged**